lib/ide/src/controller/module.rs

```diff
@@ -134,30 +142,44 @@ impl Handle {
         file_manager.touch(path.clone()).await?;
         let content = file_manager.read(path).await?;
         logger.info(|| "Parsing code");
-        let ast     = parser.parse(content,default())?;
+        let (ast,metadata) = parser.parse_file(content)?;
         logger.info(|| "Code parsed");
         logger.trace(|| format!("The parsed ast is {:?}", ast));
-        let id_map  = default();
-        let data    = Controller {location,ast,file_manager,parser,id_map,logger};
+        let id_map  = ast.id_map();
+        let data    = Controller {location,ast,file_manager,parser,id_map,metadata,logger};
         Ok(Handle::new_from_data(data))
     }
 
     /// Save the module to file.
-    pub fn save_file(&self) -> impl Future<Output=Result<(),RpcError>> {
-        let (path,mut fm) = self.with_borrowed(|data| {
-            (data.location.to_path(),data.file_manager.clone_ref())
+    pub async fn save_file(&self) -> impl Future<Output=Result<(),RpcError>> {
```

Comment:
I'd separate the file contents generaing method to a separate function. It is not specific to file saving. (e.g. we will want to pass the updated file contents to language server even when file is not "saved")